packages/flutter_driver/lib/src/common/wait.dart

```diff
@@ -0,0 +1,325 @@
+      throw SerializationException('Error occurred during deserializing the FirstFrameRasterizedCondition JSON string: $json');
+    return const FirstFrameRasterizedCondition();
+  }
+
+  @override
+  bool get condition => WidgetsBinding.instance.firstFrameRasterized;
+
+  @override
+  Future<void> wait() async {
+    await WidgetsBinding.instance.waitUntilFirstFrameRasterized;
+    assert(condition);
+  }
+
+  @override
+  Map<String, String> serialize() {
+    return <String, String>{
+      'conditionName': 'FirstFrameRasterizedCondition',
+    };
+  }
+}
+
+/// A combined condition that waits until all the given [conditions] are met.
+class CombinedCondition implements WaitCondition {
+  /// Creates a [CombinedCondition] instance with the given list of
+  /// [conditions].
+  ///
+  /// The [conditions] argument cannot be null.
+  const CombinedCondition(this.conditions)
+    : assert(conditions != null);
+
+  /// Factory constructor to parse a [CombinedCondition] instance from the given
+  /// JSON map.
+  ///
+  /// The [jsonMap] argument cannot be null.
+  factory CombinedCondition.deserialize(Map<String, dynamic> jsonMap) {
+    assert(jsonMap != null);
+    if (jsonMap['conditionName'] != 'CombinedCondition')
+      throw SerializationException('Error occurred during deserializing the CombinedCondition JSON string: $jsonMap');
+    if (jsonMap['conditions'] == null) {
+      return const CombinedCondition(<WaitCondition>[]);
+    }
+
+    final List<WaitCondition> conditions = <WaitCondition>[];
+    for (Map<String, dynamic> condition in json.decode(jsonMap['conditions'])) {
+      conditions.add(_WaitConditionDecoder.deserialize(condition));
+    }
+    return CombinedCondition(conditions);
+  }
+
+  /// A list of conditions it waits for.
+  final List<WaitCondition> conditions;
+
+  @override
+  bool get condition {
+    return conditions.every((WaitCondition condition) => condition.condition);
+  }
+
+  @override
+  Future<void> wait() async {
+    while (!condition) {
+      for (WaitCondition condition in conditions) {
+        assert (condition != null);
+        await condition.wait();
+      }
+    }
+    assert(condition);
+  }
+
+  @override
+  Map<String, String> serialize() {
+    final Map<String, String> jsonMap = <String, String>{
+      'conditionName': 'CombinedCondition'
+    };
+    final List<Map<String, String>> jsonConditions = conditions.map(
+        (WaitCondition condition) {
+          assert(condition != null);
+          return condition.serialize();
+        }).toList();
+    jsonMap['conditions'] = json.encode(jsonConditions);
+    return jsonMap;
+  }
+}
+
+/// A JSON decoder that parses JSON map to a [WaitCondition] or its subclass.
+class _WaitConditionDecoder {
+  /// Parses a [WaitCondition] or its subclass from the given [json] map.
+  ///
+  /// The [json] argument cannot be null.
+  static WaitCondition deserialize(Map<String, dynamic> json) {
```

Comment:
This could probably just be a private top-level function. No need to enclose it in a fake class.